### ISSUE TITLE
Code comments and documentation is missing

### ISSUE THREAD

**Issue by yigitgursoy** [elastic/elasticsearch-net#1883]
It is really hard to upgrade from 1.7 to 2.2 without any documentation.

What can we do to improve this?

**Comment by beruic** [comment]
I completely agree. I am trying to upgrade from NEST 1.7.1 to 2.0.4, and I am having a REALLY hard time figuring out how to fix all the attribute errors.

I'm specifically having issues converting `[ElasticProperty(OptOut = true)]` and `[ElasticProperty(Index = FieldIndexOption.NotAnalyzed)]`.

**Comment by gmarz** [comment]
Totally understand.  Getting the [new documentation](https://github.com/elastic/elasticsearch-net/tree/master/docs/contents/new) out is a priority.  In the mean time feel free to open any issues with the pain points you're hitting and we'll happily help resolve them.  Also, have a look at the [2.0 breaking changes](https://github.com/elastic/elasticsearch-net/tree/master/docs/2.0-breaking-changes).

@beruic see https://github.com/elastic/elasticsearch-net/pull/1520 - attributes are now broken out by type, so to opt out a string field you'd do:

```csharp
[String(Ignore = true)]
```

Specifying a type when you simply want to ignore a property is moot, so you could just use the `[JsonIgnore]` attribute which will also work- or maybe we should introduce our own `[Ignore]` attribute.

Similarly, to set a string field to not analyzed:

```csharp
[String(Index = FieldIndexOption.NotAnalyzed)]
```

**Comment by beruic** [comment]
Great thanks! This at least got me part of the way. However it leads to more questions:

1. Am I correct that `[Number(NumberType.Integer)]` replaces `[ElasticProperty(NumericType = NumberType.Integer, Type = FieldType.Integer)]`?
2. I assume that you refer to `[JsonIgnore]` from `Newtonsoft.JSON`?
3. Will `[BsonIgnore]` (`MongoDB.Bson.Serialization.Attributes`) work as well, or must I specify both?
4. How do I set an `ICollection` and a `long` to `FieldIndexOption.NotAnalyzed`?

I'd prefer you introduced your own `[Ignore]` attribute, as this would be much less confusing, and less prone to interfere with other functionality (say, you need to JSON encode this for something else where you need to include this).
Also, it would be nice if you prepended your attributes with something, so that it would be easy to see where this came from and avoid name clashing. E.g. `[ElasticString]` instead of `[String]` and `[ElasticIgnore]` instead of `[Ignore]`.

There's a lot of dead links in #1520 so I didn't get much out of it, other than what I got from [the blog post](https://www.elastic.co/blog/ga-release-of-nest-2-0-our-dot-net-client-for-elasticsearch).

In conclusion, I can see I'll need some documentation on the new queries in order to make it home on this one. I guess I'll just have to postpone it for now, but at least I can save my current work for later :)

**Comment by Mpdreamz** [comment]
1. Yep!
2. Yep!
3. No not out of the box (would require us to take a dep on mongodb), you can hack support for other attributes by injecting your own serializer and overriding [this method](https://github.com/elastic/elasticsearch-net/blob/36cda72a17209ac67619c9879a81e5793f2569c1/src/Nest/CommonAbstractions/SerializationBehavior/JsonNetSerializer.cs#L76)
4. `[Number]` still has an [Index](https://github.com/elastic/elasticsearch-net/blob/master/src/Nest/Mapping/Types/Core/Number/NumberAttribute.cs#L16) property. `ICollection<T>` you map to whatever `T` is (string, number object).


You can ignore using attribute mapping, using a serializer specific ignore or [in code on ConnectionSettings](https://github.com/elastic/elasticsearch-net/blob/master/src/Tests/ClientConcepts/HighLevel/Mapping/AutoMap.doc.cs#L711-L715) see also [these tests and documentation](https://github.com/elastic/elasticsearch-net/blob/master/src/Tests/ClientConcepts/HighLevel/Mapping/AutoMap.doc.cs#L668).  

This gives you ample angles to tackle this problem.

As an aside, attribute mapping only allows you to express a subset of whats possible in Elasticsearch because attribute values need to be constant. Please also look into mapping your POCO's in code when you create your index.

If you have name clashes on `StringAttribute` use `[Nest.String]`, we won't reintroduce smurf typing.

Feel free to open a new issue with specific Query questions you might run into, we are here to help :smile:

**Comment by beruic** [comment]
**4a:**
The `Index` property on `[Number]` will not accept `FieldIndexOption.NotAnalyzed`.

**4b:**
So I just do `[String(Index = FieldIndexOption.NotAnalyzed)]`, even though it is an `ICollection`?
